Mathlib/MeasureTheory/Measure/ProbabilityMeasure.lean

```diff
@@ -548,4 +561,42 @@ end ProbabilityMeasure -- namespace
 
 end map -- section
 
+section MonoidalProduct
+
+open Measure
+
+/-- The monoidal product is a measurable function from the product of probability spaces over
+``α`` and ``β`` into the type of probability spaces over ``α × β``. Lemma 4.1 of
+https://doi.org/10.1016/j.aim.2020.107239.
+-/
+theorem ProbabilityMeasure.measurable_prod {α β : Type*} [MeasurableSpace α] [MeasurableSpace β] :
+    Measurable (fun (μ : ProbabilityMeasure α × ProbabilityMeasure β)
+      ↦ μ.1.toMeasure.prod μ.2.toMeasure) := by
+  apply Measurable.measure_of_isPiSystem generateFrom_prod.symm isPiSystem_prod _ (by simp)
+  simp only [mem_image2, mem_setOf_eq, forall_exists_index, and_imp]
+  intros _ u Hu v Hv Heq
+  simp_rw [← Heq, prod_prod]
+  apply Measurable.mul
+  · exact (measurable_coe Hu).comp (measurable_subtype_coe.comp measurable_fst)
+  · exact (measurable_coe Hv).comp (measurable_subtype_coe.comp measurable_snd)
+
+/-- The monoidal product is a measurabule function from the product of finite measures over
+``α`` and ``β`` into the type of finite measures over ``α × β``. -/
+theorem FiniteMeasure.measurable_prod {α β : Type*} [MeasurableSpace α] [MeasurableSpace β] :
+    Measurable (fun (μ : FiniteMeasure α × FiniteMeasure β)
+      ↦ μ.1.toMeasure.prod μ.2.toMeasure) := by
+  apply Measurable.measure_of_isPiSystem generateFrom_prod.symm isPiSystem_prod _
+  · simp_rw [← Set.univ_prod_univ, prod_prod]
+    apply Measurable.mul
+    · exact (measurable_coe MeasurableSet.univ).comp (measurable_subtype_coe.comp measurable_fst)
+    · exact (measurable_coe MeasurableSet.univ).comp (measurable_subtype_coe.comp measurable_snd)
+  simp only [mem_image2, mem_setOf_eq, forall_exists_index, and_imp]
+  intros _ u Hu v Hv Heq
+  simp_rw [← Heq, prod_prod]
+  apply Measurable.mul
+  · exact (measurable_coe Hu).comp (measurable_subtype_coe.comp measurable_fst)
+  · exact (measurable_coe Hv).comp (measurable_subtype_coe.comp measurable_snd)
```

Comment:
In this proof, you are proving the same thing twice: once for the product of any two measurable sets, and once for univ x univ. Can you instead prove it once in a `have` subproof first, then use the result in both places?

Reply:
Done. I tried golfing the `have` proof to be shorter but was not successful.